src/harness/testcases/WINNF_FT_S_MCP_testcase.py

```diff
@@ -306,3 +310,239 @@ def test_WINNF_FT_S_MCP_1(self, config_filename):
        to 3700 MHz to protect configurable IAP-protected entities and DPAs
     """
     config = loadConfig(config_filename)
+    test_type = 'MCP'
+    # Invoke MCP test steps
+    self.executeMcpTestSteps(config, test_type)
+
+  def executeMcpTestSteps(self, config, test_type):
+    """Execute all teststeps for MCP testcase and for xPR testcases till Step22
+
+    Args:
+       config: Testcase configuration
+       test_type: A string which indicates the type of testcase to be invoked("MCP"/"XPR")
+    Return: None
+    """
+    sas_test_harness_objects = []
+    domain_proxy_objects = []
+
+    # Load conditionals
+    if config['conditionalRegistrationData']:
+      self._sas_admin.PreloadRegistrationData({
+          'registrationData': config['conditionalRegistrationData']
+      })
+
+    for domain_proxy in config['domainProxyConfigs']:
+      domain_proxy_objects.append(test_harness_objects.DomainProxy(domain_proxy['cert'],\
+                                                                   domain_proxy['key'],
+                                                                   self))
+    # Step 1 : Load DPAs
+    self._sas_admin.TriggerLoadDpas()
+
+    # STEP 2 : ESC inf rms SAS about inactive DPA
+    self._sas_admin.TriggerBulkDpaActivation({'activate': False})
+
+    # Step 3 : creates multiple SAS TH, and Load predefined FAD,CBSD
+    for test_harness in config['sasTestHarnessConfigs']:
+      # Initialize SAS Test Harness Server instance to dump FAD records
+      sas_test_harness_object = SasTestHarnessServer(test_harness['sasTestHarnessName'],
+                                                           test_harness['hostName'],
+                                                           test_harness['port'],
+                                                           test_harness['serverCert'],
+                                                           test_harness['serverKey'],
+                                                           test_harness['caCert'])
+      # Start the server
+      sas_test_harness_object.start()
+
+      # informing SAS UUT about SAS Test Harnesses
+      certificate_hash = getCertificateFingerprint(test_harness['serverCert'])
+      self._sas_admin.InjectPeerSas({'certificateHash': certificate_hash,
+                                     'url': sas_test_harness_object.getBaseUrl()})
+      sas_test_harness_objects.append(sas_test_harness_object)
+   
+    # Step 4,5 : Inject IAP protected entities into UUT
+    for iteration_content in config['iterationData']:
+      cbsd_agg_list = []
```

Comment:
Yes, we have to collect the list of CBSD records/objects which are added individually and via domain proxies. The code has been updated with 'cbsds_with_domain_proxies', a list of CBSDs added via domain proxies
'cbsd_agg_list' - a list of CBSDs added individually. 
We have to consolidate both the list and provide as input to aggregate interference model, currently we are woking on this changes. It will be reflected as part of reference models APIs integration.

Reply:
Consolidating individual CBSDs and Domain Proxy CBSDs could be a pain, yes. I'm more concerned that handling individual CBSDs will effectively require re-implementing the same thing a Domain Proxy does in the test. I'd like to propose a solution to this.

For each individual CBSD create a Domain Proxy with only that single CBSD and the cert/key as given in the CBSDs config. Add each of these to the list of Domain Proxies.

This way:
1) There is no need to re-implement the register/request grant/heartbeat behavior for individual CBSDs.
2) Support for individual CBSDs will exist implicitly in the test without any need for special consideration. This should simplify the test code a bit.

Does this make sense? and have I missed anything? As far as we can tell it should work as Domain Proxy requests look the same as a CBSD request (Except the cert/keys are different and Domain Proxies can make requests for more than one CBSD)

Reply:
Yes, got the point. For each individual CBSD, we can create a domain proxy object which in turn will take care of all the CBSD related operations.
Earlier we had a discussion with Kate on the individual CBSDs presence in MCP.1 test case. Please see below excerpt comments from PR405.
 
kate-harrison on Mar 13 
Contributor
I just realized that there's another piece missing: we need the capability for individual CBSDs (which will use their own cert/key pair) to register, get a grant, etc. So in the same way we support multiple DPs, we need to support multiple individual CBSD connections.

Sorry for catching this only at the last minute -- I literally remembered only as I was writing a reply to an unrelated email. 

jni2000 on Mar 13 • edited  
Contributor
AS per the current test spec, MCP.1 test case uses only DP test harness for CBSD’s registration and grants, so we would like to know exactly in which test step we will be using the individual CBSDs ? is there any updates on the MCP.1 test case document? Please let us know your thoughts on the comment.

kate-harrison on Mar 14 
Contributor
I believe that the intention is to use individual CBSDs in every step where "DP test harness" is mentioned: so ITS can use N DPs and M individual CBSDs in each of these steps. The test spec is unfortunately not very clear on this point; you are welcome to double-check with Masoud.

Reply:
Yes, individual CBSDs are needed (to clarify in this response "individual CBSD" means a CBSD that is not in a Domain Proxy). My point is individual CBSDs behave the same way as a Domain Proxy with a single CBSD. What I am proposing is the config has individual CBSDs as is, but the test takes each individual CBSD and make a new "domain proxy" (configured with a CBSD cert/key) to handle the behavior and stick it with the rest of the domain proxies. This will simplify the handling logic.

Example (takes the place of lines 459-462)
```python
for cbsd_record in iteration_content['cbsdRecords']:
  proxy = test_harness_objects.DomainProxy(cbsd_record['clientCert'], cbsd_record['clientKey'], self)
  proxy.registerCbsdsAndRequestGrants([cbsd_record['registrationRequest']], [cbsd_record['grantRequest']])
  domain_proxy_objects.append(proxy)
```

Note: Old version of Domain Proxy initialization used for consistency with the current code, will need to be updated.

Reply:
Updated the as suggested in the comment, created new DP object for each CBSD.